File: 2015/Day24/main.cpp
#include <fstream>
#include <functional>
#include <iostream>
#include <limits>
#include <numeric>
#include <vector>

std::vector<int> read_input(const std::string& filename) {
    std::vector<int> weights;
    std::ifstream file(filename);
    int weight;
    while (file >> weight) {
        weights.push_back(weight);
    }
    return weights;
}

long long calculate_quantum_entanglement(const std::vector<int>& group) {
    return std::accumulate(group.begin(), group.end(), 1LL, std::multiplies<long long>());
}

bool find_combination(const std::vector<int>& weights, int target_weight, int group_size, std::vector<int>& combination, int start, int current_sum) {
    if (combination.size() == group_size) {
        return current_sum == target_weight;
    }
    for (int i = start; i < weights.size(); ++i) {
        combination.push_back(weights[i]);
        if (find_combination(weights, target_weight, group_size, combination, i + 1, current_sum + weights[i])) {
            return true;
        }
        combination.pop_back();
    }
    return false;
}

long long find_ideal_quantum_entanglement(const std::vector<int>& weights, int num_groups) {
    int total_weight = std::accumulate(weights.begin(), weights.end(), 0);
    int target_weight = total_weight / num_groups;
    int n = weights.size();

    long long min_qe = std::numeric_limits<long long>::max();
    int min_group_size = n;

    for (int group_size = 1; group_size <= n / num_groups; ++group_size) {
        std::vector<int> combination;
        if (find_combination(weights, target_weight, group_size, combination, 0, 0)) {
            long long qe = calculate_quantum_entanglement(combination);
            if (qe < min_qe) {
                min_qe = qe;
            }
        }
        if (min_qe != std::numeric_limits<long long>::max()) {
            break;
        }
    }

    return min_qe;
}

int main() {
    std::vector<int> weights = read_input("input.txt");
    std::cout << "Part 1: " << find_ideal_quantum_entanglement(weights, 3) << std::endl;
    std::cout << "Part 2: " << find_ideal_quantum_entanglement(weights, 4) << std::endl;
    return 0;
}

File: 2015/Day24/main.dart
import 'dart:io';

void main() async {
  final weights = await readInput('input.txt');
  print('Part 1: ${findIdealQuantomEntanglement(weights, 3)}');
  print('Part 2: ${findIdealQuantomEntanglement(weights, 4)}');
}

int calculateQuantumEntanglement(List<int> group) {
  return group.fold(1, (acc, weight) => acc * weight);
}

bool findCombination(List<int> weights, int targetWeight, int groupSize,
    List<int> combination, int start, int currentSum) {
  if (combination.length == groupSize) return currentSum == targetWeight;
  for (int i = start; i < weights.length; i++) {
    combination.add(weights[i]);
    if (findCombination(weights, targetWeight, groupSize, combination, i + 1,
        currentSum + weights[i])) return true;
    combination.removeLast();
  }
  return false;
}

int findIdealQuantomEntanglement(List<int> weights, int numGroups) {
  final totalWeight = weights.reduce((a, b) => a + b);
  final targetWeight = totalWeight ~/ numGroups;
  final n = weights.length;

  int minQE = double.maxFinite.toInt();

  for (int groupSize = 1; groupSize <= n ~/ numGroups; ++groupSize) {
    final combination = <int>[];
    if (findCombination(weights, targetWeight, groupSize, combination, 0, 0)) {
      final qe = calculateQuantumEntanglement(combination);
      if (qe < minQE) {
        minQE = qe;
      }
    }
    if (minQE != double.maxFinite.toInt()) {
      break;
    }
  }

  return minQE;
}

Future<List<int>> readInput(String filename) async {
  return await File(filename)
      .readAsLines()
      .then((l) => l.map(int.parse).toList());
}
